utils/zerovec/src/varzerovec/components.rs

```diff
@@ -0,0 +1,243 @@
+    pub fn iter(self) -> impl Iterator<Item = &'a T::VarULE> {
+        let last = iter::from_fn(move || {
+            if self.len() > 0 {
+                let start = usizeify(self.indices[self.len() - 1]);
+                let end = self.things.len();
+                Some(unsafe { self.things.get_unchecked(start..end) })
+            } else {
+                None
+            }
+        })
+        .take(1);
+        self.indices
+            .windows(2)
+            .map(move |win| {
+                let start = usizeify(win[0]);
+                let end = usizeify(win[1]);
+                unsafe { self.things.get_unchecked(start..end) }
+            })
+            .chain(last)
+            .map(|s| unsafe { T::VarULE::from_byte_slice_unchecked(s) })
+    }
+}
+
+impl<'a, T> SliceComponents<'a, T>
+where
+    T: AsVarULE,
+    T::VarULE: Ord,
+{
+    pub fn binary_search(&self, x: &T::VarULE) -> Result<usize, usize> {
+        // This code is an absolute atrocity. This code is not a place of honor. This
+        // code is known to the State of California to cause cancer.
+        //
+        // Unfortunately, the stdlib's `binary_search*` functions can only operate on slices.
+        // We do not have a slice. We have something we can .get() and index on, but that is not
+        // a slice.
+        //
+        // The `binary_search*` functions also do not have a variant where they give you the element's
+        // index, which we could otherwise use to directly index `self`.
+        // We do have `self.indices`, but these are indices into a byte buffer, which cannot in
+        // isolation be used to recoup the logical index of the element they refer to.
+        //
+        // However, `binary_search_by()` provides references to the elements of the slice being iterated.
+        // Since the layout of Rust slices is well-defined, we can do pointer arithmetic on these references
+        // to obtain the index being used by the search.
+        //
+        // It's worth noting that the slice we choose to search is irrelevant, as long as it has the appropriate
+        // length. `self.indices` is defined to have length `self.len()`, so it is convenient to use
+        // here and does not require additional allocations.
+        //
+        // The alternative to doing this is to implement our own binary search. This is significantly less fun.
+
+        let zero_index = self.indices.as_ptr() as *const _ as usize;
+        self.indices.binary_search_by(|probe: &_| {
+            // `self.indices` is a vec of unaligned u64s, so we divide by sizeof(u64)
+            // to get the actual index
+            let index = (probe as *const _ as usize - zero_index) / mem::size_of::<u64>();
+            // safety: we know this is in bounds
+            let actual_probe = unsafe { self.get_unchecked(index) };
+            actual_probe.cmp(x)
+        })
+    }
+}
+
+pub fn get_serializable_bytes<T: AsVarULE>(elements: &[T]) -> Vec<u8> {
```

Comment:
Rather than `get_serializable_bytes` it seems like you should have `from_aligned(&[T])` that creates a proper SliceComponents from a normal slice.

Reply:
I'm not sure how that would work, since there's nothing for the `SliceComponents` to refer to? We'd have to allocate the `Vec<u8>` somewhere, this is the most Rusty way to do it -- a two step process where you first allocate the Vec and then construct the SliceComponents from a refernece.

The only other way I can think of is to use an outparam

```rust
pub fn from_aligned(&[T], buf: &mut Vec<u8>) -> Self {}
```